### ISSUE TITLE
UPP GET request returns empty array

### ISSUE THREAD

**Issue by srallen** [zooniverse/Panoptes#1920]
Calling `GET` for UPP as a project owner/collaborator or using client credentials returns an empty array for the project preferences instead of the expected project preferences resource of the user I'm trying to fetch for the project I own. This should either return a clear error or the expected resource.

**Comment by zwolf** [comment]
@camallen / @marten: The issue appears to be somewhere in the intersection of restpack and the role control system. Adding "admin: true" to the request (if the user or the owner of the client creds has permission) returns what is expected. But despite theoretically having permission, in the form of the owner of the client creds being project owner/collaborator, the "project_preferences" array is empty. The rest of the response body includes the links and metadata, though.

I don't believe that anything that has changed lately has affected this, I just figure no one's needed to do it before.

Can someone shed some light on this?

**Comment by camallen** [comment]
@zwolf best bet is to write a failing spec for the scenario @srallen has identified and then work through the `scope_for` relation results to figure out why it's not returning the correct data. I'm out tomorrow and so is marten.

**Comment by camallen** [comment]
Also @srallen what does the actual request URL look like?

**Comment by camallen** [comment]
Actually this is by design, the only user that should have access to a project preference is the user that owns it. My understanding from our previous discussions was that you would know the `user_id` and `project_id` params for the update_settings end point. You won't be able to get the resource before / after updating it as you don't own it and won't have access.

**Comment by zwolf** [comment]
That's fine, then. Gravity Spy folks should know which workflow a user is on (via their own db or the classification) and regardless, they can set it. But this call should return an error instead of an empty array, cause that's pretty confusing.

**Comment by camallen** [comment]
We have to be consistent with other api end points, some of those return empty sets if you don't have access too. I'd be happy for either or but not a mix as that fails `Principle of Least Surprise` to me.

Also we should probably just be clearer in the [docs](http://docs.panoptes.apiary.io/#reference/projectpreference/retrieve-a-single-projectpreference) about who can access and the expected result if you don't have access.

**Comment by marten** [comment]
Returning an empty set is by design. The reason behind it is a security issue: if it'd return an error like "unauthorised", we'd be leaking the information that a given user *has* done work on a given project.

Github for instance will also always return 404s even if repos technically exist but are private and you don't have access.

**Comment by marten** [comment]
For reference, here's Github's explanation: https://developer.github.com/v3/troubleshooting/#why-am-i-getting-a-404-error-on-a-repository-that-exists

**Comment by eatyourgreens** [comment]
To make things even more fun, the JSON API client swallows HTTP errors anyway, so PFE wouldn't be able to use a 404 error at the moment. https://github.com/zooniverse/json-api-client/issues/27

**Comment by srallen** [comment]
Ah, ok, I wasn't aware the empty array was by design. When I saw it, I just got confused by it.

We just need to change `save_settings()` in the python client to be a class method rather than an object method then so it can be passed the params it needs to work. See: https://github.com/zooniverse/panoptes-python-client/blob/master/panoptes_client/project_preferences.py#L35

Will close. Feel free to mark as "won't fix" 😃
